### ISSUE TITLE
Race in NamedPipeClientStream.TryConnect leads to unexpected exception

### ISSUE THREAD

**Issue by mikedn** [dotnet/corefx#22014]
The Windows version of [`TryConnect`](https://github.com/dotnet/corefx/blob/master/src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Windows.cs#L25) first calls `WaitNamedPipe` and if it returns "file not found" it tries again until the timeout expires. If `WaitNamedPipe` succeeds then the code calls `CreateNamedPipeClient` (`CreateFile`) and if that returns any error other than "pipe busy" it throws an exception. If the server dies between the call to `WaitNamedPipe` and `CreateNamedPipeClient` you'll end up with an unexpected  `FileNotFoundException`.

IMO `TryConnect` should just call `CreateNamedPipeClient` first and fall back to `WaitNamedPipe` if it returns "file not found" or "pipe busy".

**Comment by mikedn** [comment]
Sample repro:
```C#
Action s = () =>
{
    while (true)
    {
        using (var p = new NamedPipeServerStream("foo"))
        {
            try
            {
                p.WaitForConnection();
            }
            catch (IOException)
            {
            }
        }
    }
};

Action c = () =>
{
    while (true)
    {
        try
        {
            using (var p = new NamedPipeClientStream("foo"))
                p.Connect();
        }
        catch (FileNotFoundException ex)
        {
            Console.WriteLine("huh?!");
        }
    }
};

Task.WaitAll(Task.Run(s), Task.Run(c), Task.Run(c));
```

**Comment by pjanotti** [comment]
Closed by PR #24616
